core/src/cgroup.rs

```diff
@@ -1,238 +1,392 @@
-use std::time::{Duration, Instant};
 
-use anyhow::anyhow;
-use itertools::Itertools;
+use anyhow::{anyhow, bail, Ok};
+use nix::sys::statfs;
 use nix::unistd::Pid;
-use polling::{Event, Poller};
-use walkdir::WalkDir;
-
-use crate::error::{ResultExt, SystemError, TypedResult};
-
-/// An internal reference to a certain cgroup
-// TODO think about completely changing this.
-// Because CGroups are a hierarchy and Parents need to consider their children,
-// A different representation may be necessary, probably a tree with n children.
-// Also maybe we dont even need to delete the cgroups after we are done
-// We may need to verify "Domain" "Domain Threaded" and "Thread" state of CGroups
-// TODO it should be possible to use an already created cgroup incase the user provides us a group with cpuset-enabled
-#[derive(Debug, Clone)]
+
+/// A single cgroup inside our tree of managed cgroups
+///
+/// The tree is not represented by a traditional tree data structure,
+/// as this is very complicated in Rust. Instead, the tree is "calculated"
+/// by the path alone.
+#[derive(Debug)]
 pub struct CGroup {
     path: PathBuf,
 }
 
 impl CGroup {
-    const MEMBER_FILE: &'static str = "cgroup.procs";
-    const FREEZE_FILE: &'static str = "cgroup.freeze";
-    const EVENTS_FILE: &'static str = "cgroup.events";
-
-    /// Returns the first mount point on the host filesystem of the cgroup that this process belongs to
-    pub fn mount_point() -> TypedResult<PathBuf> {
-        procfs::process::Process::myself()
-            .typ(SystemError::Panic)?
-            .mountinfo()
-            .typ(SystemError::Panic)?
-            .iter()
-            .find(|m| m.fs_type.eq("cgroup2")) // TODO A process can have several cgroup mounts
-            .ok_or_else(|| anyhow!("no cgroup2 mount found"))
-            .typ(SystemError::Panic)
-            .map(|m| m.mount_point.clone())
-    }
-
-    /// Create a cgroup inside another one
-    // TODO: Switch to a tree data structure for cgroups.
-    pub fn new<P: AsRef<Path>>(parent: P, child_name: &str) -> TypedResult<Self> {
-        let path = PathBuf::from(parent.as_ref()).join(child_name);
-        //trace!("New CGroup: {path:?}");
+    /// Creates a new cgroup as the root of a sub-tree
+    ///
+    /// path must be the path of an already existing cgroup
+    pub fn new_root<P: AsRef<Path>>(path: P, name: &str) -> anyhow::Result<Self> {
+        let path = PathBuf::from(path.as_ref()).join(name);
+
+        // There is no need to check if the path already exists, as create_dir()
+        // will fail under this circumstance.
+        fs::create_dir(&path)?;
+
+        Self::import_root(&path)
+    }
+
+    /// Imports an already existing cgroup as the root of a sub-tree
+    pub fn import_root<P: AsRef<Path>>(path: P) -> anyhow::Result<Self> {
+        let path = PathBuf::from(path.as_ref());
+
+        if !is_cgroup(&path)? {
+            bail!("{} is not a valid cgroup", path.to_str().unwrap());
```

Comment:
Maybe the `unwrap` can be circumvented by using [`display`](https://doc.rust-lang.org/std/path/struct.PathBuf.html#method.display) instead.  
`bail!("{} is not a valid cgroup", path.display());`  
or a debug print
`bail!("{path:?} is not a valid cgroup");`

Reply:
See feae7d3